lib/digital_signal/digital_signal.c

```diff
@@ -72,102 +98,377 @@ uint32_t digital_signal_get_edges_cnt(DigitalSignal* signal) {
     LL_TIM_EnableUpdateEvent(TIM2);
     LL_TIM_EnableDMAReq_UPDATE(TIM2);
+}
 
-    // Start transactions
-    LL_TIM_GenerateEvent_UPDATE(TIM2); // Do we really need it?
+static void digital_signal_start_timer() {
+    LL_TIM_GenerateEvent_UPDATE(TIM2);
     LL_TIM_EnableCounter(TIM2);
+}
+
+void digital_signal_send(DigitalSignal* signal, const GpioPin* gpio) {
+    furi_assert(signal);
+
+    if(!signal->edge_cnt) {
+        return;
+    }
+
+    /* Configure gpio as output */
+    signal->gpio = gpio;
+    furi_hal_gpio_init(signal->gpio, GpioModeOutputPushPull, GpioPullNo, GpioSpeedVeryHigh);
+
+    /* single signal, add a temporary, terminating edge at the end */
+    signal->edge_timings[signal->edge_cnt++] = 10;
+    digital_signal_prepare(signal);
+
+    digital_signal_setup_dma(signal);
+    digital_signal_setup_timer();
+    digital_signal_start_timer();
+
+    while(!LL_DMA_IsActiveFlag_TC2(DMA1)) {
+    }
+
+    digital_signal_stop_timer();
+    digital_signal_stop_dma();
+
+    signal->edge_cnt--;
+}
+
+void digital_sequence_alloc_signals(DigitalSequence* sequence, uint32_t size) {
+    sequence->signals_size = size;
+    sequence->signals = malloc(sequence->signals_size * sizeof(DigitalSignal*));
+    sequence->signals_prolonged = malloc(sequence->signals_size * sizeof(bool));
+}
+
+void digital_sequence_alloc_sequence(DigitalSequence* sequence, uint32_t size) {
+    sequence->sequence_used = 0;
+    sequence->sequence_size = size;
+    sequence->sequence = malloc(sequence->sequence_size);
+}
+
+DigitalSequence* digital_sequence_alloc(uint32_t size, const GpioPin* gpio) {
+
+    DigitalSequence* sequence = malloc(sizeof(DigitalSequence));
+
+    sequence->gpio = gpio;
+    sequence->bake = false;
+
+    digital_sequence_alloc_signals(sequence, 32);
+    digital_sequence_alloc_sequence(sequence, size);
+
+    return sequence;
+}
+
+void digital_sequence_free(DigitalSequence* sequence) {
+    furi_assert(sequence);
 
-    while(!LL_DMA_IsActiveFlag_TC2(DMA1))
-        ;
+    free(sequence->signals);
+    free(sequence->sequence);
+    free(sequence);
+}
+
+void digital_sequence_set_signal(DigitalSequence* sequence, uint8_t signal_index, DigitalSignal* signal) {
+    furi_assert(sequence);
+    furi_assert(signal);
+    furi_assert(signal_index < sequence->signals_size);
+
+    sequence->signals[signal_index] = signal;
+    signal->gpio = sequence->gpio;
+    signal->reload_reg_remainder = 0;
+
+    digital_signal_prepare(signal);
+}
+
+void digital_sequence_add(DigitalSequence* sequence, uint8_t signal_index) {
+    furi_assert(sequence);
+    furi_assert(signal_index < sequence->signals_size);
+
+    if(sequence->sequence_used >= sequence->sequence_size) {
+        sequence->sequence_size += 256;
+        sequence->sequence = realloc(sequence->sequence, sequence->sequence_size);
+    }
+
+    sequence->sequence[sequence->sequence_used++] = signal_index;
+}
+
+void digital_signal_update_dma(DigitalSignal* signal) {
```

Comment:
I know this looks a bit low level.
We also can use a C-only version but then minimum length of a the *last* pulse in a signal has to be a bit longer.

Reply:
I don't think that C version will be slower here. This code is quite straight forward, no special instructions used. Please provide C-version.